### DISCUSSION TITLE
Port @eslint-community/regexpp

### DISCUSSION THREAD

**Discussion by togami2864** [biomejs/biome#1559] [Rule suggestion]
There's some ECMA regex related rules like [no-misleading-character-class](https://eslint.org/docs/latest/rules/no-misleading-character-class), [no-useless-escape](https://eslint.org/docs/latest/rules/no-useless-escape/).

ESLint leverages [@eslint-community/regexpp](https://www.npmjs.com/package/@eslint-community/regexpp), ecma script regex parer to implement these rules, while Biome lacks a dedicated regex parser, leading to the implementation of simple regex parsers within individual rules to manage diverse regex inputs.

e.g.
- https://github.com/biomejs/biome/blob/bc3bd4e18cd08c19855999247bb711644f273e14/crates/biome_js_analyze/src/semantic_analyzers/nursery/no_misleading_character_class.rs#L291-L618
- https://github.com/biomejs/biome/blob/593da224fab906fc97d71b0df137f92bb7132f87/crates/biome_js_analyze/src/analyzers/suspicious/no_control_characters_in_regex.rs#L73-L198

And some bugs have been reported.
- https://github.com/biomejs/biome/issues/1522#issue-2075898117
- https://github.com/biomejs/biome/issues/1522#issuecomment-1886955309

Consequently, I propose porting [@eslint-community/regexpp](https://www.npmjs.com/package/@eslint-community/regexpp) into biome. An additional note is made of the absence of an ECMA regex parser in Rust on crates.io.

pros:
- It will be able to provide more accurate lint
- It may prevent duplication of implementation

cons: 
- The costs may be high, while the use case remains limited.

This is just an idea. What do you think?

**Comment by Conaclos** [comment]
I think it is a good idea! Do you want to implement this parser?

**Reply by togami2864** [reply]
Sure. I'll try implement small PoC.

**Comment by Conaclos** [comment]
[OXC is also adding a regex parser](https://github.com/oxc-project/oxc/pull/2030). Could be worth to take a look at :)

**Reply by ematipico** [reply]
A regex parser is a massive endeavour. Do we have the resources? Can't we use the `regex` crate at this point?

**Reply by Conaclos** [reply]
Is this compatible with js regexes?

**Reply by togami2864** [reply]
It seems that some syntax isn't supported.
https://docs.rs/regex/1.10.2/regex/#syntax
e.g.
- lookahead
- unicode property escape
- lookbehind assertion

**Reply by Conaclos** [reply]
We could also use the oxc crate if it got merged.

**Reply by togami2864** [reply]
That sounds good too. Let's wait and see for a little while.
